dopamine/discrete_domains/run_experiment.py

```diff
@@ -555,10 +570,36 @@ def __init__(
         environment, and returns an agent.
       create_environment_fn: A function which receives a problem name and
         creates a Gym environment for that problem (e.g. an Atari 2600 game).
+      max_simultaneous_iterations: int, maximum number of iterations running
+        simultaneously in separate threads.
       **kwargs: Additional positional arguments.
     """
     threading_utils.initialize_local_attributes(
         self, _environment=create_environment_fn)
+    self._running_iterations = threading.Semaphore(max_simultaneous_iterations)
+    self._output_lock = threading.Lock()
+
     super(AsyncRunner, self).__init__(
         base_dir=base_dir, create_agent_fn=create_agent_fn,
         create_environment_fn=create_environment_fn, **kwargs)
+
+  # TODO(aarg): Decouple experience generation from training.
+  def _run_experiment_loop(self):
+    """Runs iterations in multiple threads until `num_iterations` is reached."""
+    threads = []
+    for iteration in range(self._start_iteration, self._num_iterations):
+      self._running_iterations.acquire()
+      threads.append(self._run_one_iteration(iteration))
+    # Wait for all running iterations to complete.
+    for thread in threads:
+      thread.join()
+
+  @async_method
+  def _run_one_iteration(self, iteration):
+    """Runs one iteration in separate thread, logs and checkpoints results."""
+    statistics = super(AsyncRunner, self)._run_one_iteration(iteration)
+    with self._output_lock:
```

Comment:
I'm a little confused about what's being recorded.  It's complicated by not understanding what an "iteration" represents.

Is it possible for checkpointing to be done only after training?  (and, I guess, is that a desirable or sensible thing?)

Reply:
An iteration correspond to N training steps - where N is user-specified. The iteration is going to run as many episodes as necessary until it reaches N steps. It also includes X steps of evaluation and logging / checkpointing.

You want checkpointing to be run periodically, during training (typically at each iteration).

We could checkpoint and log in a separate thread (we can do that later on). For now it makes sense to keep the existing logic and checkpoint at the end of each iteration.

Reply:
Now that I understand what an iteration encompasses (and that training is a part of it), I'm happy with how checkpointing is done.